File: shimmer/src/main/java/com/facebook/shimmer/ShimmerDrawable.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.shimmer;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.LinearGradient;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.PixelFormat;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffXfermode;
import android.graphics.RadialGradient;
import android.graphics.Rect;
import android.graphics.Shader;
import android.graphics.drawable.Drawable;
import android.view.animation.LinearInterpolator;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public final class ShimmerDrawable extends Drawable {
  private final ValueAnimator.AnimatorUpdateListener mUpdateListener =
      new ValueAnimator.AnimatorUpdateListener() {
        @Override
        public void onAnimationUpdate(ValueAnimator animation) {
          invalidateSelf();
        }
      };

  private final Paint mShimmerPaint = new Paint();
  private final Rect mDrawRect = new Rect();
  private final Matrix mShaderMatrix = new Matrix();

  private @Nullable ValueAnimator mValueAnimator;
  private float mStaticAnimationProgress = -1f;

  private @Nullable Shimmer mShimmer;

  public ShimmerDrawable() {
    mShimmerPaint.setAntiAlias(true);
  }

  public void setShimmer(@Nullable Shimmer shimmer) {
    mShimmer = shimmer;
    if (mShimmer != null) {
      mShimmerPaint.setXfermode(
          new PorterDuffXfermode(
              mShimmer.alphaShimmer ? PorterDuff.Mode.DST_IN : PorterDuff.Mode.SRC_IN));
    }
    updateShader();
    updateValueAnimator();
    invalidateSelf();
  }

  public @Nullable Shimmer getShimmer() {
    return mShimmer;
  }

  /** Starts the shimmer animation. */
  public void startShimmer() {
    startShimmer(null);
  }

  /** Starts the shimmer animation with animation listener. */
  public void startShimmer(Animator.AnimatorListener listener) {
    if (mValueAnimator != null && !isShimmerStarted() && getCallback() != null) {
      if (listener != null) {
        mValueAnimator.addListener(listener);
      }
      mValueAnimator.start();
    }
  }

  /** Stops the shimmer animation. */
  public void stopShimmer() {
    if (mValueAnimator != null && isShimmerStarted()) {
      mValueAnimator.cancel();
    }
  }

  /** Return whether the shimmer animation has been started. */
  public boolean isShimmerStarted() {
    return mValueAnimator != null && mValueAnimator.isStarted();
  }

  /** Return whether the shimmer animation is running. */
  public boolean isShimmerRunning() {
    return mValueAnimator != null && mValueAnimator.isRunning();
  }

  @Override
  public void onBoundsChange(Rect bounds) {
    super.onBoundsChange(bounds);
    mDrawRect.set(bounds);
    updateShader();
    maybeStartShimmer();
  }

  public void setStaticAnimationProgress(float value) {
    if (Float.compare(value, mStaticAnimationProgress) == 0
        || (value < 0f && mStaticAnimationProgress < 0f)) {
      return;
    }
    mStaticAnimationProgress = Math.min(value, 1f);
    invalidateSelf();
  }

  public void clearStaticAnimationProgress() {
    setStaticAnimationProgress(-1f);
  }

  @Override
  public void draw(@NonNull Canvas canvas) {
    if (mShimmer == null || mShimmerPaint.getShader() == null) {
      return;
    }

    final float tiltTan = (float) Math.tan(Math.toRadians(mShimmer.tilt));
    final float translateHeight = mDrawRect.height() + tiltTan * mDrawRect.width();
    final float translateWidth = mDrawRect.width() + tiltTan * mDrawRect.height();
    final float dx;
    final float dy;
    final float animatedValue;

    if (mStaticAnimationProgress < 0f) {
      animatedValue = mValueAnimator != null ? (float) mValueAnimator.getAnimatedValue() : 0f;
    } else {
      animatedValue = mStaticAnimationProgress;
    }

    switch (mShimmer.direction) {
      default:
      case Shimmer.Direction.LEFT_TO_RIGHT:
        dx = offset(-translateWidth, translateWidth, animatedValue);
        dy = 0;
        break;
      case Shimmer.Direction.RIGHT_TO_LEFT:
        dx = offset(translateWidth, -translateWidth, animatedValue);
        dy = 0f;
        break;
      case Shimmer.Direction.TOP_TO_BOTTOM:
        dx = 0f;
        dy = offset(-translateHeight, translateHeight, animatedValue);
        break;
      case Shimmer.Direction.BOTTOM_TO_TOP:
        dx = 0f;
        dy = offset(translateHeight, -translateHeight, animatedValue);
        break;
    }

    mShaderMatrix.reset();
    mShaderMatrix.setRotate(mShimmer.tilt, mDrawRect.width() / 2f, mDrawRect.height() / 2f);
    mShaderMatrix.preTranslate(dx, dy);
    mShimmerPaint.getShader().setLocalMatrix(mShaderMatrix);
    canvas.drawRect(mDrawRect, mShimmerPaint);
  }

  @Override
  public void setAlpha(int alpha) {
    // No-op, modify the Shimmer object you pass in instead
  }

  @Override
  public void setColorFilter(@Nullable ColorFilter colorFilter) {
    // No-op, modify the Shimmer object you pass in instead
  }

  @Override
  public int getOpacity() {
    return mShimmer != null && (mShimmer.clipToChildren || mShimmer.alphaShimmer)
        ? PixelFormat.TRANSLUCENT
        : PixelFormat.OPAQUE;
  }

  private float offset(float start, float end, float percent) {
    return start + (end - start) * percent;
  }

  private void updateValueAnimator() {
    if (mShimmer == null) {
      return;
    }

    final boolean started;
    if (mValueAnimator != null) {
      started = mValueAnimator.isStarted();
      mValueAnimator.cancel();
      mValueAnimator.removeAllUpdateListeners();
    } else {
      started = false;
    }

    mValueAnimator =
        ValueAnimator.ofFloat(0f, 1f + (float) (mShimmer.repeatDelay / mShimmer.animationDuration));
    mValueAnimator.setInterpolator(new LinearInterpolator());
    mValueAnimator.setRepeatMode(mShimmer.repeatMode);
    mValueAnimator.setStartDelay(mShimmer.startDelay);
    mValueAnimator.setRepeatCount(mShimmer.repeatCount);
    mValueAnimator.setDuration(mShimmer.animationDuration + mShimmer.repeatDelay);
    mValueAnimator.addUpdateListener(mUpdateListener);
    if (started) {
      mValueAnimator.start();
    }
  }

  void maybeStartShimmer() {
    if (mValueAnimator != null
        && !mValueAnimator.isStarted()
        && mShimmer != null
        && mShimmer.autoStart
        && getCallback() != null) {
      mValueAnimator.start();
    }
  }

  private void updateShader() {
    final Rect bounds = getBounds();
    final int boundsWidth = bounds.width();
    final int boundsHeight = bounds.height();
    if (boundsWidth == 0 || boundsHeight == 0 || mShimmer == null) {
      return;
    }
    final int width = mShimmer.width(boundsWidth);
    final int height = mShimmer.height(boundsHeight);

    final Shader shader;
    switch (mShimmer.shape) {
      default:
      case Shimmer.Shape.LINEAR:
        boolean vertical =
            mShimmer.direction == Shimmer.Direction.TOP_TO_BOTTOM
                || mShimmer.direction == Shimmer.Direction.BOTTOM_TO_TOP;
        int endX = vertical ? 0 : width;
        int endY = vertical ? height : 0;
        shader =
            new LinearGradient(
                0, 0, endX, endY, mShimmer.colors, mShimmer.positions, Shader.TileMode.CLAMP);
        break;
      case Shimmer.Shape.RADIAL:
        shader =
            new RadialGradient(
                width / 2f,
                height / 2f,
                (float) (Math.max(width, height) / Math.sqrt(2)),
                mShimmer.colors,
                mShimmer.positions,
                Shader.TileMode.CLAMP);
        break;
    }

    mShimmerPaint.setShader(shader);
  }
}

File: shimmer/src/main/java/com/facebook/shimmer/ShimmerDrawable.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.facebook.shimmer

import android.animation.ValueAnimator
import android.animation.ValueAnimator.AnimatorUpdateListener
import android.graphics.*
import android.graphics.drawable.Drawable
import android.view.animation.LinearInterpolator

class ShimmerDrawable : Drawable() {
  private val mUpdateListener = AnimatorUpdateListener { invalidateSelf() }
  private val mShimmerPaint = Paint()
  private val mDrawRect = Rect()
  private val mShaderMatrix = Matrix()
  private var mValueAnimator: ValueAnimator? = null
  private var mShimmer: Shimmer? = null
  var shimmer: Shimmer?
    get() = mShimmer
    set(shimmer) {
      mShimmer = shimmer
      if (mShimmer != null) {
        mShimmerPaint.xfermode = PorterDuffXfermode(
            if (mShimmer!!.alphaShimmer) PorterDuff.Mode.DST_IN else PorterDuff.Mode.SRC_IN)
      }
      updateShader()
      updateValueAnimator()
      invalidateSelf()
    }

  /** Starts the shimmer animation.  */
  fun startShimmer() {
    if (mValueAnimator != null && !isShimmerStarted && callback != null) {
      mValueAnimator!!.start()
    }
  }

  /** Stops the shimmer animation.  */
  fun stopShimmer() {
    if (mValueAnimator != null && isShimmerStarted) {
      mValueAnimator!!.cancel()
    }
  }

  /** Return whether the shimmer animation has been started.  */
  val isShimmerStarted: Boolean
    get() = mValueAnimator != null && mValueAnimator!!.isStarted

  public override fun onBoundsChange(bounds: Rect) {
    super.onBoundsChange(bounds)
    mDrawRect.set(bounds)
    updateShader()
    maybeStartShimmer()
  }

  override fun draw(canvas: Canvas) {
    if (mShimmer == null || mShimmerPaint.shader == null) {
      return
    }
    val tiltTan = Math.tan(Math.toRadians(mShimmer!!.tilt.toDouble())).toFloat()
    val translateHeight = mDrawRect.height() + tiltTan * mDrawRect.width()
    val translateWidth = mDrawRect.width() + tiltTan * mDrawRect.height()
    val dx: Float
    val dy: Float
    val animatedValue = if (mValueAnimator != null) mValueAnimator!!.animatedValue as Float else 0f
    when (mShimmer!!.direction) {
      Shimmer.Direction.LEFT_TO_RIGHT -> {
        dx = offset(-translateWidth, translateWidth, animatedValue)
        dy = 0f
      }
      Shimmer.Direction.RIGHT_TO_LEFT -> {
        dx = offset(translateWidth, -translateWidth, animatedValue)
        dy = 0f
      }
      Shimmer.Direction.TOP_TO_BOTTOM -> {
        dx = 0f
        dy = offset(-translateHeight, translateHeight, animatedValue)
      }
      Shimmer.Direction.BOTTOM_TO_TOP -> {
        dx = 0f
        dy = offset(translateHeight, -translateHeight, animatedValue)
      }
      else -> {
        dx = offset(-translateWidth, translateWidth, animatedValue)
        dy = 0f
      }
    }
    mShaderMatrix.reset()
    mShaderMatrix.setRotate(mShimmer!!.tilt, mDrawRect.width() / 2f, mDrawRect.height() / 2f)
    mShaderMatrix.postTranslate(dx, dy)
    mShimmerPaint.shader.setLocalMatrix(mShaderMatrix)
    canvas.drawRect(mDrawRect, mShimmerPaint)
  }

  override fun setAlpha(alpha: Int) {
    // No-op, modify the Shimmer object you pass in instead
  }

  override fun setColorFilter(colorFilter: ColorFilter?) {
    // No-op, modify the Shimmer object you pass in instead
  }

  override fun getOpacity(): Int {
    return if (mShimmer != null && (mShimmer!!.clipToChildren || mShimmer!!.alphaShimmer)) PixelFormat.TRANSLUCENT else PixelFormat.OPAQUE
  }

  private fun offset(start: Float, end: Float, percent: Float): Float {
    return start + (end - start) * percent
  }

  private fun updateValueAnimator() {
    if (mShimmer == null) {
      return
    }
    val started: Boolean
    if (mValueAnimator != null) {
      started = mValueAnimator!!.isStarted
      mValueAnimator!!.cancel()
      mValueAnimator!!.removeAllUpdateListeners()
    } else {
      started = false
    }
    mValueAnimator = ValueAnimator.ofFloat(0f, 1f + (mShimmer!!.repeatDelay / mShimmer!!.animationDuration).toFloat())
    mValueAnimator!!.interpolator = LinearInterpolator()
    mValueAnimator!!.repeatMode = mShimmer!!.repeatMode
    mValueAnimator!!.startDelay = mShimmer!!.startDelay
    mValueAnimator!!.repeatCount = mShimmer!!.repeatCount
    mValueAnimator!!.duration = mShimmer!!.animationDuration + mShimmer!!.repeatDelay
    mValueAnimator!!.addUpdateListener(mUpdateListener)
    if (started) {
      mValueAnimator!!.start()
    }
  }

  fun maybeStartShimmer() {
    if (mValueAnimator != null && !mValueAnimator!!.isStarted
        && mShimmer != null && mShimmer!!.autoStart
        && callback != null) {
      mValueAnimator!!.start()
    }
  }

  private fun updateShader() {
    val bounds = bounds
    val boundsWidth = bounds.width()
    val boundsHeight = bounds.height()
    if (boundsWidth == 0 || boundsHeight == 0 || mShimmer == null) {
      return
    }
    val width = mShimmer!!.width(boundsWidth.toFloat())
    val height = mShimmer!!.height(boundsHeight.toFloat())
    val shader: Shader
    when (mShimmer!!.shape) {
      Shimmer.Shape.LINEAR -> {
        val vertical = (mShimmer!!.direction == Shimmer.Direction.TOP_TO_BOTTOM
            || mShimmer!!.direction == Shimmer.Direction.BOTTOM_TO_TOP)
        val endX = if (vertical) 0F else width
        val endY = if (vertical) height else 0F
        shader = LinearGradient(
            0F, 0F, endX, endY, mShimmer!!.colors, mShimmer!!.positions, Shader.TileMode.CLAMP)
      }
      Shimmer.Shape.RADIAL -> shader = RadialGradient(
          width / 2f,
          height / 2f,
          (Math.max(width, height) / Math.sqrt(2.0)).toFloat(),
          mShimmer!!.colors,
          mShimmer!!.positions,
          Shader.TileMode.CLAMP)
      else -> {
        val vertical = (mShimmer!!.direction == Shimmer.Direction.TOP_TO_BOTTOM
            || mShimmer!!.direction == Shimmer.Direction.BOTTOM_TO_TOP)
        val endX = if (vertical) 0F else width
        val endY = if (vertical) height else 0F
        shader = LinearGradient(
            0F, 0F, endX, endY, mShimmer!!.colors, mShimmer!!.positions, Shader.TileMode.CLAMP)
      }
    }
    mShimmerPaint.shader = shader
  }

  init {
    mShimmerPaint.isAntiAlias = true
  }
}